types/firestore.d.ts

```diff
@@ -50,6 +50,37 @@ declare namespace FirebaseFirestore {
    */
   function setLogFunction(logger: ((msg: string) => void) | null): void;
 
+  // export type DataWithFieldValue<T> = {[P in keyof T]: T[P] | FieldValue};
+  //
+  type WithFieldValue<T> = T extends number
+    ? T | NumericFieldValue
+    : T extends Timestamp
+    ? T | TimestampFieldValue
+    : T extends Array<infer U>
+    ? T | ArrayFieldValue
+    : T;
+
+  export type Primitive =
+    | string
+    | number
+    | boolean
+    | bigint
+    | symbol
+    | undefined
+    | null;
+  export type Builtin = Primitive | Function;
+
+  /** Like Partial but recursive */
+  export type NestedPartial<T> = T extends Builtin
+    ? T
+    : T extends Map<infer K, infer V>
+    ? Map<NestedPartial<K>, NestedPartial<V>>
+    : T extends Array<infer U>
+    ? Array<NestedPartial<U>>
+    : T extends {}
+    ? {[K in keyof T]?: NestedPartial<T[K] | WithFieldValue<T[K]>>}
```

Comment:
`WithFieldValue<T[K]>` includes `T[K]` already.
```suggestion
    ? {[K in keyof T]?: NestedPartial<WithFieldValue<T[K]>>}
```
Do you want to accept partial `FieldValue`s? It feels a little weird, but probably doesn't matter because no one will pass partial `FieldValue`s in practice. You could add a branch for `FieldValue`s and return them as is like the `Builtin`s. Up to you.